src/math/ring_signature.cpp

```diff
@@ -0,0 +1,363 @@
+        // Insert into map
+        keys.insert({public_key, secret});
+    }
+    return keys;
+}
+
+// Make a list of public keys for which we have the corresponding secret key
+// in a single ring of public keys.
+const point_list known_keys_in_ring(
+    const secret_keys_map& secret_keys,
+    const point_list& ring)
+{
+    point_list known_ring;
+    for (const auto& key: ring)
+    {
+        // Is this public key in our secret keys map?
+        if (secret_keys.find(key) != secret_keys.end())
+            known_ring.push_back(key);
+    }
+    return known_ring;
+}
+
+// For all rings, make a list of known public keys corresponding to each ring.
+const key_rings partition_keys_into_rings(
+    const secret_keys_map& secret_keys,
+    const key_rings& rings)
+{
+    key_rings known_keys;
+    // For each ring in our rings...
+    for (const auto& ring: rings)
+    {
+        // ... find known keys corresponding to our private keys
+        const auto known_ring = known_keys_in_ring(secret_keys, ring);
+        // And add this list to our rings of known public keys.
+        known_keys.push_back(known_ring);
+    }
+    return known_keys;
+}
+
+// Check that all rings are not empty
+bool all_rings_nonzero_size(const key_rings& known_keys)
+{
+    for (const auto& ring: known_keys)
+        if (ring.empty())
+            return false;
+    return true;
+}
+
+typedef std::vector<size_t> index_list;
+
+// Make a list of indexes of where our known key occurs in each ring of
+// public keys. That is given a ring of {A, B, C} where we know the
+// private key of B, it will return 1 (the index in the ring).
+// This function computes this for all rings.
+index_list search_key_indexes(
+    const key_rings& rings,
+    const key_rings& known_keys_by_ring)
+{
+    index_list known_key_indexes;
+    // Iterate all our rings.
+    for (size_t i = 0; i < rings.size(); ++i)
+    {
+        BITCOIN_ASSERT(i < known_keys_by_ring.size());
+        const auto& ring = rings[i];
+        const auto& known = known_keys_by_ring[i];
+        // Find known key in this current ring.
+        const auto it = std::find(ring.begin(), ring.end(), known.back());
+        BITCOIN_ASSERT(it != ring.end());
+        // Calculate and store the index
+        size_t index = it - ring.begin();
+        known_key_indexes.push_back(index);
+    }
+    return known_key_indexes;
+}
+
+// Returns message || flatten(rings)
+data_chunk concatenate(const data_slice message, const key_rings& rings)
+{
+    data_chunk result(message.begin(), message.end());
+    for (const auto& ring: rings)
+        for (const auto& key: ring)
+            extend_data(result, key);
+    return result;
+}
+
+// Generate random k and s values.
+// There is a single k for each ring.
+// Each key in each ring, has a corresponding s value.
+// That is, each ring[i][j] should have a corresponding s[i][j]
+void random_k_and_s_values(
+    secret_list& k, ring_signature::s_values_type& s,
```

Comment:
Prefer `out_k` and `out_s` here.

Reply:
Style: don't wrap parameters until hit 80 char limit.

Reply:
ok done